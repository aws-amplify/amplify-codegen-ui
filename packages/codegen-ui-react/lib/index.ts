/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
export * from './react-studio-template-renderer-helper';
export * from './react-component-renderer';
export * from './react-expander-renderer';
export * from './react-table-renderer';
export * from './imports';
export * from './react-studio-template-renderer';
export * from './react-theme-studio-template-renderer';
export * from './react-output-config';
export * from './react-render-config';
export * from './react-output-manager';
export * from './amplify-ui-renderers/amplify-renderer';
export * from './amplify-ui-renderers/amplify-form-renderer';
export * from './amplify-ui-renderers/amplify-view-renderer';
export * from './primitive';
export * from './react-index-studio-template-renderer';
export * from './react-utils-studio-template-renderer';
export * from './react-required-dependency-provider';
export * from './utils-file-functions';
