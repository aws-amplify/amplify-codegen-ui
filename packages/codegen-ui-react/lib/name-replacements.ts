/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

// List of character and description source https://www.ascii-code.com/
const nameReplacements: [RegExp, string][] = [
  [/ /g, ''], // remove spaces
  [/!/g, 'ExclamationMark'],
  [/"/g, 'DoubleQuotes'],
  [/#/g, 'Nmber'],
  [/\$/g, 'Dollar'],
  [/&/g, 'Ampersand'],
  [/'/g, 'SingleQuote'],
  [/\(/g, 'OpenParenthesis'],
  [/\)/g, 'CloseParenthesis'],
  [/\*/g, 'Asterisk'],
  [/\+/g, 'Plus'],
  [/,/g, 'Comma'],
  [/-/g, ''], // remove hypens
  [/\./g, 'Period'],
  [/\//g, 'Slash'],
  [/0/g, 'Zero'],
  [/1/g, 'One'],
  [/2/g, 'Two'],
  [/3/g, 'Three'],
  [/4/g, 'Four'],
  [/5/g, 'Five'],
  [/6/g, 'Six'],
  [/7/g, 'Seven'],
  [/8/g, 'Eight'],
  [/9/g, 'Nine'],
  [/:/g, 'Colon'],
  [/;/g, 'Semicolon'],
  [/</g, 'LessThan'],
  [/=/g, 'Equal'],
  [/>/g, 'GreaterThan'],
  [/\?/g, 'QuestionMark'],
  [/@/g, 'AtSymbol'],
  [/\[/g, 'OpeningBracket'],
  [/\]/g, 'ClosingBracket'],
  [/\^/g, 'Caret'],
  [/_/g, 'UnderScore'], // remove underscores
  [/`/g, 'GraveAccent'],
  [/~/g, 'Tilde'],
];

export default nameReplacements;
