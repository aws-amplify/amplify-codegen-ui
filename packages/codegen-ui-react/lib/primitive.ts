/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { FixedStudioComponentProperty } from '@aws-amplify/codegen-ui';
import { factory, SyntaxKind, TypeParameterDeclaration, TypeNode } from 'typescript';

export type PrimitiveLevelPropConfiguration<ConfigType> = {
  [componentType: string]: { [eventType: string]: ConfigType };
};

export enum Primitive {
  Alert = 'Alert',
  Badge = 'Badge',
  Button = 'Button',
  ButtonGroup = 'ButtonGroup',
  Card = 'Card',
  CheckboxField = 'CheckboxField',
  Collection = 'Collection',
  Divider = 'Divider',
  Expander = 'Expander',
  ExpanderItem = 'ExpanderItem',
  Flex = 'Flex',
  Grid = 'Grid',
  Heading = 'Heading',
  Icon = 'Icon',
  Image = 'Image',
  Link = 'Link',
  Loader = 'Loader',
  Menu = 'Menu',
  MenuButton = 'MenuButton',
  MenuItem = 'MenuItem',
  Pagination = 'Pagination',
  PasswordField = 'PasswordField',
  PhoneNumberField = 'PhoneNumberField',
  Placeholder = 'Placeholder',
  Radio = 'Radio',
  RadioGroupField = 'RadioGroupField',
  Rating = 'Rating',
  ScrollView = 'ScrollView',
  SearchField = 'SearchField',
  SelectField = 'SelectField',
  SliderField = 'SliderField',
  StepperField = 'StepperField',
  SwitchField = 'SwitchField',
  Table = 'Table',
  TableBody = 'TableBody',
  TableCell = 'TableCell',
  TableFoot = 'TableFoot',
  TableHead = 'TableHead',
  TableRow = 'TableRow',
  Tabs = 'Tabs',
  TabItem = 'TabItem',
  Text = 'Text',
  TextAreaField = 'TextAreaField',
  TextField = 'TextField',
  ToggleButton = 'ToggleButton',
  ToggleButtonGroup = 'ToggleButtonGroup',
  View = 'View',
  VisuallyHidden = 'VisuallyHidden',
  Autocomplete = 'Autocomplete',
}

/*
 * temporary list of Primitives with a change event. Final implementation will pull from amplify UI
 */
export const PrimitivesWithChangeEvent: Set<Primitive> = new Set([
  Primitive.CheckboxField,
  Primitive.PasswordField,
  Primitive.PhoneNumberField,
  Primitive.RadioGroupField,
  Primitive.SearchField,
  Primitive.SelectField,
  Primitive.SliderField,
  Primitive.StepperField,
  Primitive.SwitchField,
  Primitive.TextAreaField,
  Primitive.TextField,
]);

export const PrimitiveChildrenPropMapping: Partial<Record<Primitive, string>> = {
  [Primitive.Alert]: 'label',
  [Primitive.Badge]: 'label',
  [Primitive.Button]: 'label',
  [Primitive.Heading]: 'label',
  [Primitive.Link]: 'label',
  [Primitive.MenuButton]: 'label',
  [Primitive.MenuItem]: 'label',
  [Primitive.Radio]: 'label',
  [Primitive.TableCell]: 'label',
  [Primitive.Text]: 'label',
  [Primitive.ToggleButton]: 'label',
};

export const PrimitiveDefaultPropertyValue: PrimitiveLevelPropConfiguration<FixedStudioComponentProperty> = {
  [Primitive.CheckboxField]: {
    checked: { value: false, type: 'boolean' },
  },
  [Primitive.PasswordField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.PhoneNumberField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.RadioGroupField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.SearchField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.SwitchField]: {
    isChecked: { value: false, type: 'boolean' },
  },
  [Primitive.TextAreaField]: {
    value: { value: '', type: 'string' },
  },
  [Primitive.TextField]: {
    value: { value: '', type: 'string' },
  },
};

export function isPrimitive(componentType: string): boolean {
  return Object.values(Primitive).includes(componentType as Primitive);
}

export const PrimitiveTypeParameter: Partial<
  Record<Primitive, { declaration: () => TypeParameterDeclaration[] | undefined; reference: () => TypeNode[] }>
> = {
  [Primitive.TextField]: {
    declaration: () => [
      factory.createTypeParameterDeclaration(
        factory.createIdentifier('Multiline'),
        factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
        undefined,
      ),
    ],
    reference: () => [factory.createTypeReferenceNode(factory.createIdentifier('Multiline'), undefined)],
  },
  [Primitive.Collection]: {
    declaration: () => undefined,
    reference: () => [factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)],
  },
};

export const PRIMITIVE_OVERRIDE_PROPS = 'PrimitiveOverrideProps';
/**
 * export declare type PrimitiveOverrideProp<T> = Partial<T> & React.DOMAttributes<HTMLDivElement>;
 */
export const primitiveOverrideProp = factory.createTypeAliasDeclaration(
  undefined,
  [factory.createModifier(SyntaxKind.ExportKeyword), factory.createModifier(SyntaxKind.DeclareKeyword)],
  factory.createIdentifier(PRIMITIVE_OVERRIDE_PROPS),
  [factory.createTypeParameterDeclaration(factory.createIdentifier('T'), undefined, undefined)],
  factory.createIntersectionTypeNode([
    factory.createTypeReferenceNode(factory.createIdentifier('Partial'), [
      factory.createTypeReferenceNode(factory.createIdentifier('T'), undefined),
    ]),
    factory.createTypeReferenceNode(
      factory.createQualifiedName(factory.createIdentifier('React'), factory.createIdentifier('DOMAttributes')),
      [factory.createTypeReferenceNode(factory.createIdentifier('HTMLDivElement'), undefined)],
    ),
  ]),
);
