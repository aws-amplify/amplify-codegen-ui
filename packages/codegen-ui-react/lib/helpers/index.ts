/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { GenericDataField, GenericDataSchema, StudioFormFields } from '@aws-amplify/codegen-ui/lib/types';
import { factory, Statement, Expression, NodeFlags, Identifier, ParameterDeclaration, SyntaxKind } from 'typescript';
import { isPrimitive } from '../primitive';

export const lowerCaseFirst = (input: string) => input.charAt(0).toLowerCase() + input.slice(1);

export const createUniqueName = (name: string, isNameUsed: (input: string) => boolean) => {
  if (!isNameUsed(name)) {
    return name;
  }
  let count = 0;
  const prospectiveNewName = name;
  while (isNameUsed(prospectiveNewName + count)) {
    count += 1;
  }
  return prospectiveNewName + count;
};

export const capitalizeFirstLetter = (val: string) => {
  return val.charAt(0).toUpperCase() + val.slice(1);
};

export const getSetNameIdentifier = (value: string): Identifier => {
  return factory.createIdentifier(`set${capitalizeFirstLetter(value)}`);
};

export const getModelNameProp = (value: string): string => {
  return `${lowerCaseFirst(value)}ModelProp`;
};

/**
 * const [name, setName] = React.useState({default_expression});
 *
 * name is the value we are looking to set
 * defaultValue is is the value to set for the useState
 * @param name
 * @param defaultValue
 * @returns
 */
export const buildUseStateExpression = (name: string, defaultValue: Expression): Statement => {
  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createArrayBindingPattern([
            factory.createBindingElement(undefined, undefined, factory.createIdentifier(name), undefined),
            factory.createBindingElement(undefined, undefined, getSetNameIdentifier(name), undefined),
          ]),
          undefined,
          undefined,
          factory.createCallExpression(
            factory.createPropertyAccessExpression(
              factory.createIdentifier('React'),
              factory.createIdentifier('useState'),
            ),
            undefined,
            [defaultValue],
          ),
        ),
      ],
      NodeFlags.Const,
    ),
  );
};

export const buildUseRefExpression = (name: string, defaultValue?: Expression): Statement => {
  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(name),
          undefined,
          undefined,
          factory.createCallExpression(
            factory.createPropertyAccessExpression(
              factory.createIdentifier('React'),
              factory.createIdentifier('useRef'),
            ),
            undefined,
            defaultValue ? [defaultValue] : undefined,
          ),
        ),
      ],
      NodeFlags.Const,
    ),
  );
};

/**
 * Create statement to declare and initialized a const.
 *
 * const name = value;
 * @param name
 * @param value
 * @returns
 */
export const buildInitConstVariableExpression = (name: string, value: Expression): Statement => {
  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [factory.createVariableDeclaration(factory.createIdentifier(name), undefined, undefined, value)],
      NodeFlags.Const,
    ),
  );
};

/**
 * Create statement to declare an arrow function to a const.
 *
 * @example
 * ```
 * const variableName = (parameterDeclaration) => {
 *  functionName(expression);
 * };
 * ```
 */
export const buildArrowFunctionStatement = (
  variableName: string,
  functionName: string,
  expression: Expression,
  parameterDeclarations?: ParameterDeclaration[],
): Statement => {
  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(variableName),
          undefined,
          undefined,
          factory.createArrowFunction(
            undefined,
            undefined,
            parameterDeclarations ?? [],
            undefined,
            factory.createToken(SyntaxKind.EqualsGreaterThanToken),
            factory.createBlock(
              [
                factory.createExpressionStatement(
                  factory.createCallExpression(factory.createIdentifier(functionName), undefined, [expression]),
                ),
              ],
              true,
            ),
          ),
        ),
      ],
      NodeFlags.Const,
    ),
  );
};

export function fieldNeedsRelationshipLoadedForCollection(field: GenericDataField, dataSchema: GenericDataSchema) {
  const { relationship, dataType } = field;
  if (!relationship || !dataType) {
    return false;
  }
  if (typeof dataType !== 'object' || !('model' in dataType)) {
    return false;
  }
  const joinTableNames = Object.entries(dataSchema.models).reduce((acc, [modelName, { isJoinTable }]) => {
    if (isJoinTable) {
      acc.add(modelName);
    }
    return acc;
  }, new Set());
  // return false if manyToMany
  if (joinTableNames.has(dataType.model)) {
    return false;
  }
  return true;
}

export function modelNeedsRelationshipsLoadedForCollection(modelName: string, dataSchema?: GenericDataSchema): boolean {
  return !!(
    dataSchema &&
    dataSchema.models[modelName] &&
    Object.values(dataSchema.models[modelName].fields).some((field) =>
      fieldNeedsRelationshipLoadedForCollection(field, dataSchema),
    )
  );
}

export function isAliased(componentType: string): boolean {
  return !!(isPrimitive(removeAlias(componentType)) && componentType.match(/(Custom$)/g));
}

export function removeAlias(aliasString: string): string {
  return aliasString.replace(/(Custom$)/g, '');
}

export function getControlledComponentDefaultValue(
  fields: StudioFormFields,
  componentType: string,
  name: string,
): string | null {
  let defaultValue = null;
  Object.entries(fields).forEach(([key, value]) => {
    if (key === name && 'inputType' in value && value.inputType?.defaultValue && componentType === 'TextField') {
      defaultValue = value.inputType.defaultValue;
    }
  });
  return defaultValue;
}
