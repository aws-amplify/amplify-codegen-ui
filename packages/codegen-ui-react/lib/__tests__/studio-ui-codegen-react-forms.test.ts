/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { StudioTemplateRendererFactory } from '@aws-amplify/codegen-ui/lib/template-renderer-factory';
import { DataStoreModelInfo, StudioForm, SchemaModel } from '@aws-amplify/codegen-ui';
import { ReactRenderConfig, AmplifyFormRenderer, ModuleKind, ScriptKind, ScriptTarget } from '..';
import { loadSchemaFromJSONFile } from './__utils__';

export const defaultCLIRenderConfig: ReactRenderConfig = {
  module: ModuleKind.ES2020,
  target: ScriptTarget.ES2020,
  script: ScriptKind.JSX,
  renderTypeDeclarations: true,
};

export const generateWithAmplifyFormRenderer = (
  formJsonFile: string,
  dataSchemaJsonFile: string | undefined,
  renderConfig: ReactRenderConfig = defaultCLIRenderConfig,
): { componentText: string; declaration?: string } => {
  let dataStoreInfo: DataStoreModelInfo | undefined;
  if (dataSchemaJsonFile) {
    const { fields } = loadSchemaFromJSONFile<SchemaModel>(dataSchemaJsonFile);
    dataStoreInfo = { fields: Object.values(fields).map((value) => value) };
  }
  const rendererFactory = new StudioTemplateRendererFactory(
    (component: StudioForm) => new AmplifyFormRenderer(component, dataStoreInfo, renderConfig),
  );

  const renderer = rendererFactory.buildRenderer(loadSchemaFromJSONFile<StudioForm>(formJsonFile));
  return renderer.renderComponent();
};

describe('amplify form renderer tests', () => {
  describe('datastore form tests', () => {
    it('should generate a create form', () => {
      const { componentText, declaration } = generateWithAmplifyFormRenderer(
        'forms/post-datastore-create',
        'datastore/post',
      );
      expect(componentText).toContain('useDataStoreCreateAction');
      expect(componentText).toMatchSnapshot();
      expect(declaration).toMatchSnapshot();
    });

    it('should generate a update form', () => {
      const { componentText, declaration } = generateWithAmplifyFormRenderer(
        'forms/post-datastore-update',
        'datastore/post',
      );
      expect(componentText).toContain('useDataStoreUpdateAction');
      expect(componentText).toMatchSnapshot();
      expect(declaration).toMatchSnapshot();
    });
  });

  describe('custom form tests', () => {
    it('should render a custom backed form', () => {
      const { componentText, declaration } = generateWithAmplifyFormRenderer('forms/post-custom-create', undefined);
      expect(componentText).toContain('customDataFormOnSubmit(customDataFormFields)');
      expect(componentText).toMatchSnapshot();
      expect(declaration).toMatchSnapshot();
    });
  });
});
