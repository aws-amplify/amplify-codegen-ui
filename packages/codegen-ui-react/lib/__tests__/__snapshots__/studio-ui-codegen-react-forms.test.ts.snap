// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify form renderer tests custom form tests should render a custom backed form 1`] = `
"/* eslint-disable */
import React from \\"react\\";
import {
  getOverrideProps,
  useStateMutationAction,
} from \\"@aws-amplify/ui-react/internal\\";
import { Button, Flex, Grid, TextField } from \\"@aws-amplify/ui-react\\";
export default function customDataForm(props) {
  const {
    onSubmit: customDataFormOnSubmit,
    onCancel,
    overrides,
    ...rest
  } = props;
  const [customDataFormFields, setCustomDataFormFields] =
    useStateMutationAction({});
  return (
    <form
      onSubmit={async (event) => {
        event.preventDefault();
        customDataFormOnSubmit(customDataFormFields);
      }}
      {...rest}
      {...getOverrideProps(overrides, \\"customDataForm\\")}
    >
      <Grid
        columnGap=\\"15px\\"
        rowGap=\\"15px\\"
        {...getOverrideProps(overrides, \\"customDataFormGrid\\")}
      >
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid0\\")}
        >
          <TextField
            label=\\"name\\"
            isRequired={true}
            defaultValue=\\"John Doe\\"
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid1\\")}
        >
          <TextField
            label=\\"E-mail\\"
            isRequired={true}
            defaultValue=\\"johndoe@amplify.com\\"
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
      </Grid>
      <Flex
        justifyContent=\\"space-between\\"
        marginTop=\\"1rem\\"
        {...getOverrideProps(overrides, \\"CTAFlex\\")}
      >
        <Button
          label=\\"Cancel\\"
          type=\\"button\\"
          {...getOverrideProps(overrides, \\"CancelButton\\")}
        ></Button>
        <Flex {...getOverrideProps(overrides, \\"SubmitAndResetFlex\\")}>
          <Button
            label=\\"Clear\\"
            type=\\"reset\\"
            {...getOverrideProps(overrides, \\"ClearButton\\")}
          ></Button>
          <Button
            label=\\"Submit\\"
            type=\\"submit\\"
            variation=\\"primary\\"
            {...getOverrideProps(overrides, \\"SubmitButton\\")}
          ></Button>
        </Flex>
      </Flex>
    </form>
  );
}
"
`;

exports[`amplify form renderer tests custom form tests should render a custom backed form 2`] = `
"import React from \\"react\\";
import { EscapeHatchProps } from \\"@aws-amplify/ui-react/internal\\";
export declare type customDataFormProps = React.PropsWithChildren<{
    overrides?: EscapeHatchProps | undefined | null;
} & {
    onSubmit: (fields: Record<string, string>) => void;
    onCancel?: () => void;
}>;
export default function customDataForm(props: customDataFormProps): React.ReactElement;
"
`;

exports[`amplify form renderer tests datastore form tests should generate a create form 1`] = `
"/* eslint-disable */
import React from \\"react\\";
import {
  getOverrideProps,
  useDataStoreCreateAction,
  useStateMutationAction,
} from \\"@aws-amplify/ui-react/internal\\";
import { Post } from \\"../models\\";
import { schema } from \\"../models/schema\\";
import { Button, Flex, Grid, TextField } from \\"@aws-amplify/ui-react\\";
export default function myPostForm(props) {
  const { onSubmitBefore, onSubmitComplete, onCancel, overrides, ...rest } =
    props;
  const [myPostFormFields, setMyPostFormFields] = useStateMutationAction({});
  const myPostFormOnSubmit = useDataStoreCreateAction({
    model: Post,
    fields: myPostFormFields,
    schema: schema,
  });
  return (
    <form
      onSubmit={async (event) => {
        event.preventDefault();
        myPostFormOnSubmit();
      }}
      {...rest}
      {...getOverrideProps(overrides, \\"myPostForm\\")}
    >
      <Grid
        columnGap=\\"15px\\"
        rowGap=\\"15px\\"
        {...getOverrideProps(overrides, \\"myPostFormGrid\\")}
      >
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid0\\")}
        >
          <TextField
            label=\\"caption\\"
            isRequired={false}
            isReadOnly={false}
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid1\\")}
        >
          <TextField
            label=\\"username\\"
            isRequired={false}
            isReadOnly={false}
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid2\\")}
        >
          <TextField
            label=\\"post_url\\"
            isRequired={false}
            isReadOnly={false}
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid3\\")}
        >
          <TextField
            label=\\"profile_url\\"
            isRequired={false}
            isReadOnly={false}
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
      </Grid>
      <Flex
        justifyContent=\\"space-between\\"
        marginTop=\\"1rem\\"
        {...getOverrideProps(overrides, \\"CTAFlex\\")}
      >
        <Button
          label=\\"Cancel\\"
          type=\\"button\\"
          {...getOverrideProps(overrides, \\"CancelButton\\")}
        ></Button>
        <Flex {...getOverrideProps(overrides, \\"SubmitAndResetFlex\\")}>
          <Button
            label=\\"Clear\\"
            type=\\"reset\\"
            {...getOverrideProps(overrides, \\"ClearButton\\")}
          ></Button>
          <Button
            label=\\"Submit\\"
            type=\\"submit\\"
            variation=\\"primary\\"
            {...getOverrideProps(overrides, \\"SubmitButton\\")}
          ></Button>
        </Flex>
      </Flex>
    </form>
  );
}
"
`;

exports[`amplify form renderer tests datastore form tests should generate a create form 2`] = `
"import React from \\"react\\";
import { EscapeHatchProps } from \\"@aws-amplify/ui-react/internal\\";
export declare type myPostFormProps = React.PropsWithChildren<{
    overrides?: EscapeHatchProps | undefined | null;
} & {
    onSubmitBefore?: (fields: Record<string, string>) => Record<string, string>;
    onSubmitComplete?: ({ saveSuccessful, errorMessage }: {
        saveSuccessful: string;
        errorMessage?: string;
    }) => void;
    onCancel?: () => void;
}>;
export default function myPostForm(props: myPostFormProps): React.ReactElement;
"
`;

exports[`amplify form renderer tests datastore form tests should generate a update form 1`] = `
"/* eslint-disable */
import React from \\"react\\";
import {
  getOverrideProps,
  useDataStoreUpdateAction,
  useStateMutationAction,
} from \\"@aws-amplify/ui-react/internal\\";
import { Post } from \\"../models\\";
import { schema } from \\"../models/schema\\";
import { Button, Flex, Grid, TextField } from \\"@aws-amplify/ui-react\\";
export default function myPostForm(props) {
  const { id, onSubmitBefore, onSubmitComplete, onCancel, overrides, ...rest } =
    props;
  const [myPostFormFields, setMyPostFormFields] = useStateMutationAction({});
  const myPostFormOnSubmit = useDataStoreUpdateAction({
    model: Post,
    fields: myPostFormFields,
    id: id,
    schema: schema,
  });
  return (
    <form
      onSubmit={async (event) => {
        event.preventDefault();
        myPostFormOnSubmit();
      }}
      {...rest}
      {...getOverrideProps(overrides, \\"myPostForm\\")}
    >
      <Grid
        columnGap=\\"15px\\"
        rowGap=\\"15px\\"
        {...getOverrideProps(overrides, \\"myPostFormGrid\\")}
      >
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid0\\")}
        >
          <TextField
            label=\\"caption\\"
            isRequired={false}
            isReadOnly={false}
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid1\\")}
        >
          <TextField
            label=\\"username\\"
            isRequired={false}
            isReadOnly={false}
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid2\\")}
        >
          <TextField
            label=\\"post_url\\"
            isRequired={false}
            isReadOnly={false}
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
        <Grid
          columnGap=\\"inherit\\"
          rowGap=\\"inherit\\"
          templateColumns=\\"repeat(1, auto)\\"
          {...getOverrideProps(overrides, \\"RowGrid3\\")}
        >
          <TextField
            label=\\"profile_url\\"
            isRequired={false}
            isReadOnly={false}
            {...getOverrideProps(overrides, \\"TextField0\\")}
          ></TextField>
        </Grid>
      </Grid>
      <Flex
        justifyContent=\\"space-between\\"
        marginTop=\\"1rem\\"
        {...getOverrideProps(overrides, \\"CTAFlex\\")}
      >
        <Button
          label=\\"Cancel\\"
          type=\\"button\\"
          {...getOverrideProps(overrides, \\"CancelButton\\")}
        ></Button>
        <Flex {...getOverrideProps(overrides, \\"SubmitAndResetFlex\\")}>
          <Button
            label=\\"Clear\\"
            type=\\"reset\\"
            {...getOverrideProps(overrides, \\"ClearButton\\")}
          ></Button>
          <Button
            label=\\"Submit\\"
            type=\\"submit\\"
            variation=\\"primary\\"
            {...getOverrideProps(overrides, \\"SubmitButton\\")}
          ></Button>
        </Flex>
      </Flex>
    </form>
  );
}
"
`;

exports[`amplify form renderer tests datastore form tests should generate a update form 2`] = `
"import React from \\"react\\";
import { EscapeHatchProps } from \\"@aws-amplify/ui-react/internal\\";
export declare type myPostFormProps = React.PropsWithChildren<{
    overrides?: EscapeHatchProps | undefined | null;
} & {
    id: string;
    onSubmitBefore?: (fields: Record<string, string>) => Record<string, string>;
    onSubmitComplete?: ({ saveSuccessful, errorMessage }: {
        saveSuccessful: string;
        errorMessage?: string;
    }) => void;
    onCancel?: () => void;
}>;
export default function myPostForm(props: myPostFormProps): React.ReactElement;
"
`;
